### PR TITLE
131 implement bicameral query/fetch for daymet:

### PR DESCRIPTION
**Pull request by ags-tolson**
TLDR:  I think I have a way for new code to have natural return values, yet still fulfill the convention that that requires a list and a dict.

New implementation of superclass query_service does the list/dict
wrapping that the DH expects.  Clean up a few unusual things
formatting-wise and add comments and docstrings.  Remove URL & asset
filename construction to query_provider, a class method.

### CODE REVIEW

File: gips/data/daymet/daymet.py [hunk 1]
```diff
@@ -133,24 +139,35 @@ def __init__(self, filename):
...
+        """Determine availability of data for the given (asset, tile, date).
+
+        Returns (basename, url) on success; (None, None) otherewise.
+        Something of a no-op for daymet as the data is daily for the
+        entire run of the dataset and the URLs are deterministic.
+        """
+        source = cls._assets[asset]['source']
+        url = (cls._assets[asset]['url'] + '/%s') % (date.year, tile, date.year, source)
+        bn = cls._basename_pat.format(asset, tile, date.year, date.strftime('%j'))
+        return (bn, url)
```
Comment (ircwaves) [inline gips/data/daymet/daymet.py]: whether double-`None` or list, it should not return info for non-fetchable data:
```
In [10]: daymetAsset.query_provider('tmin', '9404', date(1979, 5, 25))
Out[10]: 
('daymet_tmin_9404_1972146.tif',
 'https://thredds.daac.ornl.gov/thredds/dodsC/ornldaac/1328/tiles/1979/9404_1979/tmin.nc')
```
bounds checking on `cls.start_date(asset)` and `cls.end_date(asset)` is in order.

File: gips/data/daymet/daymet.py [hunk 2]
```diff
@@ -133,24 +139,35 @@ def __init__(self, filename):
...
+    @classmethod
     def fetch(cls, asset, tile, date):
-        """ Get this asset for this tile and date (using OpenDap service) """
-        url = cls._assets[asset].get('url', '') % (date.year, tile, date.year)
-        source = cls._assets[asset]['source'] 
-        loc = "%s/%s" % (url, source)
-        print loc
-        dataset = open_url(loc)
+        """Fetch a daymet asset and convert it to a gips-friendly format."""
+        asset_bn, url = cls.query_provider(asset, tile, date)
```
Comment (ircwaves) [inline gips/data/daymet/daymet.py]: Should this check for Nones somewhere?  I believe the `gips` move is to return w/o error if data cannot be fetched for that time-space request.  This looks a lot like it hits a `TypeError` or a `404` when it tries to fetch bad dates, which will get caught in the error_handler above and deliver confusing errors.

File: gips/data/core.py [hunk 3]
```diff
@@ -357,6 +357,18 @@ def dates(cls, asset, tile, dates, days):
...
+
+    @classmethod
+    def query_service(cls, asset, tile, date):
+        """Wrap call to emulate existing convention."""
+        if not hasattr(cls, 'query_provider'):
+            raise NotImplementedError('Query/fetch bifurcation not supported for ' + cls.__name__)
+        bn, url = cls.query_provider(asset, tile, date)
+        if (bn, url) == (None, None):
+            return []
+        return [{'basename': bn, 'url': url}]
```
Comment (ircwaves) [inline gips/data/core.py]: Add documentation along the lines of:

"""
Asset child classes may implement the `query_service` in 
any manner they see fit, so long as they return a `list` of `dict`s
for each file which must be fetched to provide the requested
`(asset, tile, date)`, which may be an empty list.

Some drivers defer this to a `query_provider` method which 
must return a couple of `(basename, url)`, or couple of `(None, None)`.
"""

File: gips/data/daymet/daymet.py (hunk 2 above)
Comment (ags-tolson) [inline gips/data/daymet/daymet.py]: It should once the date-checking is added to `query_provider`; right now it can't return Nones I don't think.

**Comment by ags-tolson** [comment]
requested changes + dev merge are in place, but testing revealed some problems with core end_date; should have something pushed before standup

**Pull request merged**